sycl/doc/SharedLibraries.md

```diff
@@ -0,0 +1,276 @@
+# Shared DPC++ libraries
+
+This document describes purpose and design of Shared DPC++ libraries feature.
+
+## Background
+Sometimes users want to provide *device* functions via shared libraries.
+Simple source example:
+```
+// App:
+
+CGH.parallel_for<app_kernel>(/* ... */ {
+  library_function();
+});
+
+
+// Shared library:
+SYCL_EXTERNAL void library_function() {
+  // do something
+}
+```
+It is possible to manually create `sycl::program` in both app and shared
+library, then use `link` SYCL API to get a single program and launch kernels
+using it. But it is not user-friendly and it is very different from regular
+C/C++ workflow.
+
+The main purpose of this feature is to provide a mechanism which allows to
+provide *device* functions via shared libraries and works as close as possible
+to regular shared libraries.
+
+## Requrements:
+User's code is compiled into a shared library which consists of some host API,
+device code and device API (`SYCL_EXTERNAL` functions). The library is linked to
+a user's application which also contains some device code and performs
+computations using DPC++/SYCL.
+For this combination the following statements must be true:
+
+- `SYCL_EXTERNAL` functions from library can be called (directly or indirectly)
+  from device code of the application.
+- Function pointers taken in application should work inside the library.
+- Specific code changes are not required, i.e. the mechanism of linking works
+  as close as possible to regular shared libraries.
+
+## Design
+The overall idea is simple:
+
+- Each device image is supplied with an information about exported and imported
+  symbols using device image properties
+- DPC++ RT performs *device images collection* task by grouping all device
+  images required to execute a kernel based on the list of exports/imports
+  - Besides symbol names, additional attributes are taken into account (like
+    device image format: SPIR-V or device asm)
+- Actual linking is performed by underlying backend (OpenCL/L0/etc.)
+
+Next sections describe details of changes in each component.
```

Comment:
I suggest to extend the intro a bit for better clarity. Could you please describe here (on abstract level) how main components work together:
- what additional information is attached to device images - basically, what information export/import lists hold
- logical binding between host module and export/import lists (if any - is it dropped now?)
- how runtime performs resolution of dependencies based on that info.

Reply:
I described what information is attached - basically it is lists of function names.

> logical binding between host module and export/import lists (if any - is it dropped now?)

Could you please clarify what did you mean here?

Reply:
There can be multiple .so with multiple device binary images in each. If device image imports `foo` - what is the algorithm of searching for the device binary containing `foo` definition? AFAIU from later doc, all available binary images' export lists are searched in no particular order, and the first match is returned. If so, please mention this in the overall idea. (if so, this would mean there is no logical binding between host module and export/import lists, and resolution is performed w/o regard to containing host modules)

Reply:
Ok, will do.